File: src/Graphics/GraphicsList.java
package Graphics;

import Models.Measurement;

import java.util.Vector;

/**
 * Stores lists of drawing instructions for lines and rectangles.
 */
public class GraphicsList {
    private final Vector<GraphicsInstructions> instructions;

    /**
     * Create an empty GraphicsList
     */
    public GraphicsList() {
        this.instructions = new Vector<>();
    }

    /**
     *
     * @return A Vector of Vectors of Instructions. The first element is for lines. The second element is for rectangles
     */
    public Vector<Vector<Vector<Integer>>> drawingInstructions() {
        Vector<Vector<Vector<Integer>>> result = new Vector<>();
        Vector<Vector<Integer>> lines = new Vector<>();
        Vector<Vector<Integer>> rectangles = new Vector<>();
        for (GraphicsInstructions graphic : this.instructions) {
            Class<? extends GraphicsInstructions> graphicClass = graphic.getClass();
            Vector<Integer> instructions = graphic.drawingInstructions();
            if (graphicClass == LineInstructions.class) {
                lines.add(instructions);
            }
            else if (graphicClass == RectangleInstructions.class) {
                rectangles.add(instructions);
            }
        }
        result.add(lines);
        result.add(rectangles);
        return result;
    }

    /**
     *
     * @param graphic - add a graphic to this list
     * @return this list
     */
    public GraphicsList addGraphic(GraphicsInstructions graphic) {
        this.instructions.add(graphic);
        return this;
    }

    /**
     *
     * @param drawingInstructions - A GraphicsList to merge with this one
     * @return this GraphicsList
     */
    public GraphicsList addGraphics(GraphicsList drawingInstructions){
        this.instructions.addAll(drawingInstructions.instructions);
        return this;
    }

    /**
     *
     * @param horizontal - how far to shift horizontally
     * @param vertical - how far to shift vertically
     * @return this GraphicsList
     */
    public GraphicsList shift(Measurement horizontal, Measurement vertical) {
        this.instructions.forEach(graphicsInstructions -> graphicsInstructions.shift(horizontal, vertical));
        return this;
    }
}

File: src/main/kotlin/graphics/GraphicsList.kt
package graphics

import models.Measurement
import java.util.*
import kotlin.collections.HashMap

/**
 * Stores lists of drawing instructions for lines and rectangles.
 */
class GraphicsList {
    private val instructions: Vector<GraphicsInstructions> = Vector()

    /**
     *
     * @return A Vector of Vectors of Instructions. The first element is for lines. The second element is for rectangles
     */
    fun drawingInstructions(): HashMap<String, Vector<Vector<Int>>> {
        val result = HashMap<String, Vector<Vector<Int>>>()
        val lines = Vector<Vector<Int>>()
        val rectangles = Vector<Vector<Int>>()
        for (graphic in this.instructions) {
            val instructions = graphic.drawingInstructions()
            if (graphic is LineInstructions) {
                lines.add(instructions)
            } else if (graphic is RectangleInstructions) {
                rectangles.add(instructions)
            }
        }
        if (!lines.isEmpty()) {
            result["lines"] = lines
        }
        if (!rectangles.isEmpty()) {
            result["rectangles"] = rectangles
        }
        return result
    }

    /**
     *
     * @param graphic - add a graphic to this list
     * @return this list
     */
    fun addGraphic(graphic: GraphicsInstructions): GraphicsList {
        instructions.add(graphic)
        return this
    }

    /**
     *
     * @param drawingInstructions - A GraphicsList to merge with this one
     * @return this GraphicsList
     */
    fun addGraphics(drawingInstructions: GraphicsList): GraphicsList {
        instructions.addAll(drawingInstructions.instructions)
        return this
    }

    /**
     *
     * @param horizontal - how far to shift horizontally
     * @param vertical - how far to shift vertically
     * @return this GraphicsList
     */
    fun shift(horizontal: Measurement, vertical: Measurement): GraphicsList {
        val result = GraphicsList()
        instructions.forEach() { result.addGraphic(it.shift(horizontal, vertical)) }
        return result
    }
}